proposals/NNNN-improve-starting-streaming-services.md

```diff
@@ -0,0 +1,54 @@
+* Status: **Awaiting review**
+* Impacted Platforms: [ iOS / Java Suite / JavaScript Suite ]
+
+## Introduction
+
+In the process of analyzing mobile navigation streaming issues, we have identified several issues and unaligned implementations between SDL Core and the libraries. This proposal addresses difficulties to handle failed `StartService` attempts for video and audio services including
+1. synchronizing `StartService` requests
+2. reset SSL connection upon failures
+3. retry mechanism for starting services
+
+## Motivation
+
+With current SDL implementation, if the app sends `StartService` with `encrypted=1` and Core's `SSLMode` is set to `CLIENT`,  Core starts a TLS handshake to authenticate the app which is required to send encrypted data on the service session. If this service is successfully started, Core will remember the handshake for the current transport connection. If Core receives a second `StartService` request with `encrypted=1` it will not restart another TLS handshake. However, if the first StartService failed, Core will reset the SSL connection and restart a TLS handshake for the next `StartService` request.
+
+The video and audio streaming managers of the SDL libraries try to start their service sessions concurrently which is a challenge to synchronize with the security manager's SSL connection. The security manager can be seen as a single resource where it's access need to be synchronized. Properly recovering from issues when the system fails to establish secured service sessions is very difficult. The SSL connection cannot be restarted safely if another `StartService` request was send.
+
+The service can fail for many different reasons. Possible failure reasons can be
+- a missing permission due to missing policy update,
+- invalid certificates pending to be updated or
+- discrepancy in clock time causing handshake to fail.
+
+With the procedure of starting services being synchronized and the ability to reset SSL connections, a retry mechanism can be established to improve the user experience launching apps especially for mobile navigation.
+
+## Proposed solution
+
+The solution is to synchronize attempts to start services so that only one starting service is processing. This is essential for services which require encryption/protection as those initiate a TLS handshake on the app's transport session. If the library should start a protected service, it has to ensure the responsible security manager is started. If the attempt to establish a protected service failed due to a `StartServiceNAK`, the library must reset the SSL connection of the security manager instance. This can be done by stopping the security manager. It should be automatically restarted upon the next `StartService` request. This ensures a fluent operation of the security manager across any protected service (including RPC). The next request must not be send within two seconds after the previous attempt failed to avoid unnecessary load for the IVI.
```

Comment:
```suggestion
The solution is to synchronize attempts to start services so that only one starting service is processing. This is essential for services which require encryption/protection as those initiate a TLS handshake on the app's transport session. If the library should start a protected service, it has to ensure the responsible security manager is started. If the attempt to establish a protected service failed due to a `StartServiceNAK`, the library must reset the SSL connection of the security manager instance. This can be done by stopping the security manager. It should be automatically restarted upon the next `StartService` request. This ensures a fluent operation of the security manager across any protected service (including RPC). The next request must not be sent within two seconds after the previous attempt failed to avoid unnecessary load for the IVI.
```